sensor/src/interfaces/mqtt.py

```diff
@@ -1,8 +1,45 @@
+import json
 from typing import Any
 from src import types
+from paho.mqtt.client import Client, MQTTMessage
+import ssl
+import queue
+
+# TODO: statically type messages
+mqtt_message_queue = queue.Queue(maxsize=1024)  # type:ignore
+
+
+# print message, useful for checking if it was successful
+def on_message(client: Client, userdata: Any, msg: MQTTMessage) -> None:
+    try:
+        payload = json.loads(msg.payload.decode())
+        mqtt_message_queue.put({"topic": msg.topic, "qos": msg.qos, "payload": payload})
```

Comment:
Using a queue here seems weird. _paho-mqtt_ uses a queue internally already if I'm not mistaken. Could you explain to me a bit more, why you don't want to handle the messages the moment they arrive?

If you want to loop manually over the messages, you can use `client.loop()` instead of `client.loop_start()`. See [here for an explanation of the paho-mqtt loop](http://www.steves-internet-guide.com/loop-python-mqtt-client/). Then you should also be able to leave the message handling logic inside the `on_message` callback.

Reply:
I want to handle the messages at a certain point in the mainloop and in the main thread. The program loops over a few steps in an infinite loop (pump, measurements, system checks, etc.). For example, I don't want a config-change-routing or software-update-routing to interrupt the main thread while the pump is running with the calibration bottles open.

Yes, I could call `client.loop()` instead of `client.loop_start()` but the incoming messages will still be handled in a separate thread. I don't want that new thread to call anything important. The hierarchy should be clear: the mainloop iterates over a set of steps.

Does that make sense to you?

Reply:
Yes, I could call `client.loop()` instead of `client.loop_start()` but the incoming messages will still be handled by the callback and have to be passed to the location where `client.loop()` would be called. There are a few options (See "Returning Values from Callbacks" section at http://www.steves-internet-guide.com/mqtt-python-callbacks/ - they use a `queue` too). Hence, it doesn't make a difference whether the queue gets filled asynchronously or synchronously.

I don't want `on_message` callback to call anything important because then error handling and hierarchy becomes a mess.

Does that make sense?

Reply:
More or less, you don't want the messages to be handled inside the paho-mqtt listener thread, is that it? paho-mqtt's callbacks make this separation a bit awkward, but I think I see what you're going for.

Reply:
Yes, exactly. I could do something like:

```python
client.on_message = on_message

def on_message(message):
    if message.payload.type == "...":
        do_something()
```

But calling do_something() here is a bit weird. Ideal would be to do:

```python
def get_messages():
    return client.loop(0.5)
```